<a href="https://colab.research.google.com/github/carlos-alves-one/-MSc-Data-Science-AI-Thesis/blob/main/historical_weather_MLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Goldsmiths University of London
### MSc. Data Science and Artificial Intelligence
### Module: Final Project in Data Science (2023-2024)
### Author: Carlos Manuel De Oliveira Alves
### Student: cdeol003

**Description:** This script performs data analysis and MLP on historical weather data for solar energy forecasting.

> This script imports the time module, records the script's start time, and prints a formatted timestamp indicating when the script began execution.

In [1]:
import time

start_time = time.time()
start_time_str = time.strftime('%Y-%m-%d %H:%M:%S')
print(f"Script started at: {start_time_str}")


Script started at: 2024-08-23 22:03:25


# Data Collection and Description

## 1. Load the Data
   - Connect to Google Drive to access the dataset
   - Load the data from the provided CSV file.

In [2]:
# Imports the 'drive' module from 'google.colab' and mounts the Google Drive to
# the '/content/drive' directory in the Colab environment.
from google.colab import drive

# This function mounts Google Drive
def mount_google_drive():
    drive.mount('/content/drive')

# Call the function to mount Google Drive
mount_google_drive()

# Import the pandas library and give it the alias 'pd' for data manipulation and analysis
import pandas as pd

# Load the dataset Amazon Review Details from Google Drive
data_path = '/content/drive/MyDrive/big_data_project/historical_weather.csv'
data = pd.read_csv(data_path)

# Display the first few rows of the dataframe
data.head(5).T


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


0                    1  \
datetime                2021-09-01 00:00:00  2021-09-01 00:00:00   
temperature                            14.2                 13.9   
dewpoint                               11.6                 11.5   
rain                                    0.0                  0.0   
snowfall                                0.0                  0.0   
surface_pressure                     1015.9               1010.7   
cloudcover_total                         31                   33   
cloudcover_low                           31                   37   
cloudcover_mid                            0                    0   
cloudcover_high                          11                    0   
windspeed_10m                      7.083333             5.111111   
winddirection_10m                         8                  359   
shortwave_radiation                     0.0                  0.0   
direct_solar_radiation                  0.0                  0.0   
diffuse_radiation                       0.0                  0.0   
latitude                               57.6                 57.6   
longitude                              21.7                 22.2   
data_block_id                           1.0                  1.0   

                                          2                    3  \
datetime                2021-09-01 00:00:00  2021-09-01 00:00:00   
temperature                            14.0                 14.6   
dewpoint                               12.5                 11.5   
rain                                    0.0                  0.0   
snowfall                                0.0                  0.0   
surface_pressure                     1015.0               1017.3   
cloudcover_total                         31                    0   
cloudcover_low                           34                    0   
cloudcover_mid                            0                    0   
cloudcover_high                           0                    0   
windspeed_10m                      6.333333             8.083333   
winddirection_10m                       355                  297   
shortwave_radiation                     0.0                358.0   
direct_solar_radiation                  0.0                277.0   
diffuse_radiation                       0.0                 81.0   
latitude                               57.6                 57.6   
longitude                              22.7                 23.2   
data_block_id                           1.0                  1.0   

                                          4  
datetime                2021-09-01 00:00:00  
temperature                            15.7  
dewpoint                               12.9  
rain                                    0.0  
snowfall                                0.0  
surface_pressure                     1014.0  
cloudcover_total                         22  
cloudcover_low                           25  
cloudcover_mid                            0  
cloudcover_high                           0  
windspeed_10m                      8.416667  
winddirection_10m                         5  
shortwave_radiation                     0.0  
direct_solar_radiation                  0.0  
diffuse_radiation                       0.0  
latitude                               57.6  
longitude                              23.7  
data_block_id                           1.0

> The dataset contains various weather variables that will be useful for your project on short-term solar energy forecasting. Here are the columns present in the dataset:

1. datetime: Timestamp of the data.
2. temperature: Temperature in degrees Celsius.
3. dewpoint: Dewpoint in degrees Celsius.
4. rain: Rainfall amount in mm.
5. snowfall: Snowfall amount in mm.
6. surface_pressure: Surface pressure in hPa.
7. cloudcover_total: Total cloud cover percentage.
8. cloudcover_low: Low-level cloud cover percentage.
9. cloudcover_mid: Mid-level cloud cover percentage.
10. cloudcover_high: High-level cloud cover percentage.
11. windspeed_10m: Wind speed at 10 meters in m/s.
12. winddirection_10m: Wind direction at 10 meters in degrees.
13. shortwave_radiation: Shortwave radiation in W/m².
14. direct_solar_radiation: Direct solar radiation in W/m².
15. diffuse_radiation: Diffuse radiation in W/m².
16. latitude: Latitude coordinate.
17. longitude: Longitude coordinate.
18. data_block_id: Data block identifier.

# Exploratory Data Analysis (EDA)

In [3]:
# Print the number of rows and columns
print("\n-> Number of rows.....: {:,}".format(data.shape[0]))
print("-> Number of columns..:", data.shape[1])

# Display a summary of the dataset
print("\n-> Summary of the dataset:\n")
print(data.describe().T)

# Check data types and missing values
print("\n-> Data types and missing values:\n")
print(data.info())

# Checking for missing values
print("\n-> Missing values in each column:\n")
print(data.isnull().sum())



-> Number of rows.....: 1,710,802
-> Number of columns..: 18

-> Summary of the dataset:

                            count         mean         std    min     25%  \
temperature             1710802.0     5.740968    8.025647  -23.7     0.0   
dewpoint                1710802.0     2.240312    7.224357  -25.9    -2.6   
rain                    1710802.0     0.049620    0.207911    0.0     0.0   
snowfall                1710802.0     0.016049    0.074629    0.0     0.0   
surface_pressure        1710802.0  1009.281515   13.088915  942.9  1001.5   
cloudcover_total        1710802.0    60.912696   37.769048    0.0    25.0   
cloudcover_low          1710802.0    46.685927   40.747598    0.0     3.0   
cloudcover_mid          1710802.0    34.406980   38.327693    0.0     0.0   
cloudcover_high         1710802.0    36.051408   41.358521    0.0     0.0   
windspeed_10m           1710802.0     4.849871    2.475450    0.0     3.0   
winddirection_10m       1710802.0   197.869419   89.937978    

> Key findings from the dataset analysis:

1. Dataset Size:
   - Number of rows: 1,710,802
   - Number of columns: 18

2. Data Types:
   - 12 columns are of type float64
   - 5 columns are of type int64
   - 1 column (datetime) is of type object

3. Missing Values:
   - There are no missing values in any of the columns

4. Key Statistics:
   a. Temperature:
      - Mean: 5.74°C
      - Minimum: -23.7°C
      - Maximum: 32.6°C

   b. Precipitation:
      - Mean rainfall: 0.05 mm
      - Maximum rainfall: 16.8 mm
      - Mean snowfall: 0.016 mm
      - Maximum snowfall: 2.66 mm

   c. Pressure:
      - Mean surface pressure: 1009.28 hPa
      - Range: 942.9 hPa to 1049.3 hPa

   d. Cloud Cover:
      - Mean total cloud cover: 60.91%
      - All cloud cover types (total, low, mid, high) range from 0% to 100%

   e. Wind:
      - Mean wind speed at 10m: 4.85 m/s
      - Maximum wind speed at 10m: 21.75 m/s
      - Wind direction ranges from 0° to 360°

   f. Solar Radiation:
      - Mean shortwave radiation: 106.49 W/m²
      - Maximum shortwave radiation: 849 W/m²

   g. Location:
      - Latitude range: 57.6° to 59.7°
      - Longitude range: 21.7° to 28.2°

5. Additional Observations:
   Given the latitude and longitude ranges, the dataset appears to cover a specific geographic area, possibly in northern Europe.
   - A wide range of weather conditions is represented, from very cold (-23.7°C) to warm (32.6°C) temperatures.
   - The data includes various meteorological parameters, making it suitable for comprehensive weather analysis.
   - The 'data_block_id' column suggests the data might be organized into different blocks or periods.

This dataset provides a rich source of meteorological information, covering temperature, precipitation, atmospheric pressure, cloud cover, wind, and solar radiation. It is a comprehensive weather dataset for the specified geographic area, with no missing values, which is excellent for analysis.

> This comprehensive meteorological dataset contains 1,710,802 rows and 18 columns with no missing values. It covers a specific geographic area, likely in northern Europe (latitude 57.6° to 59.7°, longitude 21.7° to 28.2°), and includes a wide range of weather parameters. The data shows diverse conditions, with temperatures ranging from -23.7°C to 32.6°C (mean 5.74°C), precipitation levels up to 16.8mm for rain and 2.66mm for snow, and surface pressure between 942.9 and 1049.3 hPa. Cloud cover, wind speed (up to 21.75 m/s at 10m height), wind direction, and solar radiation measurements are also included. The dataset's completeness and variety of meteorological factors make it an excellent resource for in-depth weather analysis and modelling for the region.


##Visualize Feature Distributions

In [ ]:
# Importing visualization libraries and numeric operations for data analysis and plotting
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import math

# Select only numeric columns
numeric_columns = data.select_dtypes(include=['float64', 'int64']).columns

# Determine the number of rows and columns for the subplot grid
num_plots = len(numeric_columns)
num_columns = 4
num_rows = math.ceil(num_plots / num_columns)

# Create a grid of histograms
plt.figure(figsize=(20, num_rows * 4))
for i, column in enumerate(numeric_columns, 1):
    plt.subplot(num_rows, num_columns, i)
    sns.histplot(data[column], kde=True)
    plt.tight_layout()

# Show the plots
plt.show()


> Key findings from the weather dataset visualizations:

1. Data Distribution:
   - Shows that the data is evenly distributed across different data_block_ids, with each block containing around 14,000 entries. This suggests a consistent data collection process over time or across different locations.

2. Weather Parameters:
   a. Temperature and Dewpoint:
      - Both temperature and dewpoint follow a roughly normal distribution.
      - Temperature ranges from about -20°C to 30°C, with a peak around 5-10°C.
      - Dewpoint has a similar range but peaks slightly lower, around 0-5°C.

   b. Precipitation:
      - Both rain and snowfall show extremely right-skewed distributions.
      - Most observations have little to no precipitation, with rare heavy rain or snow occurrences.

   c. Surface Pressure:
      - Follows a normal distribution, centered around 1010-1015 hPa.

   d. Cloud Cover:
      - Total, low, and mid-level cloud cover all show U-shaped distributions.
      - There are peaks at 0% (clear skies) and 100% (overcast), with fewer observations.

   e. Wind:
      - Wind speed follows a right-skewed distribution, with most speeds below ten m/s and a peak around 4-5 m/s.
      - Wind direction shows peaks at cardinal and intercardinal directions, suggesting possible measurement bias or prevalent wind patterns.

   f. Solar Radiation:
      - Shortwave, direct solar, and diffuse radiation all show right-skewed distributions, peaking at 0.
      - This is expected due to the day-night cycle and varying cloud cover.

3. Geographical Distribution:
   - Latitude and longitude plots show distinct peaks, indicating that data was collected from specific locations rather than continuously across a region.

4. Data Quality:
   - The consistent counts across data blocks and the absence of unexpected gaps or outliers in most plots suggest good overall data quality.
   - The wind direction plot shows some potential measurement biases that need further investigation.

5. Climate Insights:
   - The temperature and precipitation patterns suggest a temperate climate with occasional snow, consistent with the latitude range (around 57.5°-59.5°N).
   - The pressure distribution and cloud cover patterns indicate a location with variable weather conditions, possibly influenced by maritime factors.

These visualizations provide a comprehensive overview of the weather patterns in the dataset, highlighting the distributions and relationships between various meteorological parameters. The data is high quality and suitable for detailed climate analysis or weather modelling for the specific region it represents.

> The weather dataset visualizations reveal a comprehensive and consistently collected set of meteorological data, likely from a temperate climate region. The data is evenly distributed across data blocks, containing about 14,000 entries. Temperature and dewpoint follow normal distributions centred around 5-10°C and 0-5°C, respectively, while precipitation shows rare heavy rain or snow occurrences. Surface pressure is usually distributed around 1010-1015 hPa. Cloud cover exhibits U-shaped distributions, indicating frequent clear or overcast conditions. Wind speeds are typically below ten m/s, peaking at 4-5 m/s, with directional data showing potential measurement biases. Solar radiation data reflects expected day-night cycles and cloud cover variations. The latitude (57.5°-59.5°N) and longitude plots suggest data collection from specific locations rather than a continuous region. Overall, the visualizations indicate high-quality data suitable for detailed climate analysis, representing a temperate zone with variable weather conditions, possibly influenced by maritime factors.

##Correlation Matrix

In [ ]:
# Select only columns with numeric data types for the correlation matrix
numeric_data = data.select_dtypes(include=['float64', 'int64'])

# Display a correlation matrix
correlation_matrix = numeric_data.corr()
plt.figure(figsize=(10, 10))
sns.heatmap(correlation_matrix, annot=True, fmt='.2f', cmap='coolwarm')
plt.title('Correlation Matrix')
plt.show()


> Key findings from the correlation matrix:

1. Temperature and Dewpoint:
   - Robust positive correlation (0.93), indicating that as the temperature rises, dewpoint typically rises.

2. Cloud Cover:
   - Strong positive correlations among different cloud cover types (total, low, mid, high), with the strongest between total and low cloud cover (0.84).
   - Negative correlation with temperature, especially for low cloud cover (-0.35).

3. Solar Radiation:
   - Positive correlations among shortwave, direct solar, and diffuse radiation (0.83-0.97).
   - Moderate negative correlation with temperature (-0.40 to -0.45), suggesting cooler temperatures during higher solar radiation periods (possibly due to seasonal effects).

4. Precipitation:
   - Weak positive correlation between rain and cloud cover (0.15-0.27).
   - Snowfall shows a weak negative correlation with temperature (-0.19) and a weak positive with cloud cover (0.21-0.32).

5. Surface Pressure:
   - Weak to moderate negative correlations with cloud cover (-0.28 to -0.34) and precipitation (-0.17 to -0.26).

6. Wind:
   - Weak correlations overall, with the strongest being a negative correlation between wind speed and longitude (-0.32).

7. Geographical Factors:
   - Latitude and longitude show very weak correlations with most variables, suggesting minimal impact of location within the study area on weather patterns.

8. Data Block ID:
   - Very weak correlations with all variables indicate consistent data collection across blocks.

Overall, the matrix reveals expected weather relationships (e.g., temperature-dewpoint, cloud cover-radiation) while showing minimal geographical influences within the study area. The data appears consistent across collection periods, supporting its reliability for further analysis.

> The correlation matrix reveals several critical relationships in the weather dataset. Temperature and dewpoint show a robust positive correlation (0.93). At the same time, different types of cloud cover (total, low, mid, high) are strongly interconnected, with total and low cloud cover having the highest correlation (0.84). Solar radiation components (shortwave, direct, diffuse) are highly correlated (0.83-0.97) and show a moderate negative relationship with temperature (-0.40 to -0.45), possibly due to seasonal effects. Precipitation variables have weak to moderate correlations with cloud cover and temperature, while surface pressure shows weak negative correlations with cloud cover and precipitation. Wind parameters and geographical factors (latitude, longitude) generally exhibit weak correlations with other variables, suggesting minimal impact of location within the study area on weather patterns. The consistent weak correlations of the data block ID with all variables indicate reliable data collection across different periods. Overall, the matrix confirms expected meteorological relationships while highlighting the dataset's consistency and suitability for comprehensive weather analysis.

##Features Correlation Pair Plot

In [ ]:
# Select the variables of interest based on the important correlations
selected_vars = [
    'temperature', 'dewpoint', 'cloudcover_low', 'cloudcover_total',
    'direct_solar_radiation', 'diffuse_radiation', 'shortwave_radiation',
    'surface_pressure', 'cloudcover_mid', 'windspeed_10m', 'winddirection_10m'
]

# Sample the data for the selected variables
sampled_data = numeric_data[selected_vars].sample(frac=0.1, random_state=1)

# Create the pair plot
sns.pairplot(sampled_data)
plt.suptitle('Pair Plot of Important Weather Variable Correlations\n', fontsize=16)
plt.show()


> Key findings from the scatter plot matrix:

1. Temperature and Dewpoint: A strong positive linear relationship confirms their high correlation.

2. Precipitation (Rain and Snowfall): Both show a concentration of data points near zero, with occasional higher values, indicating infrequent precipitation events.

3. Cloud Cover (Total, Low, Mid, High): These variables show complex, non-linear relationships with each other and other parameters. There are noticeable clusters at 0% and 100% cover.

4. Solar Radiation (Shortwave, Direct, Diffuse): These show strong positive relationships with each other and a negative relationship with cloud cover.

5. Surface Pressure: It shows weak to moderate relationships with other variables, with some interesting patterns visible.

6. Wind Speed has weak relationships with most variables but shows some patterns with pressure and radiation.

7. Wind Direction: Shows a reasonably uniform distribution across all angles (0-360°), with some preference for specific directions.

8. Latitude and Longitude: Show distinct clusters, suggesting data collection from specific locations rather than a continuous area.

9. Data Block ID: Shows little relationship with other variables, indicating consistent data collection across different periods or locations.

10. Complex Interactions: Many plots show complex, non-linear relationships between variables, highlighting the intricate nature of weather systems.

This visualization reinforces the correlations seen in the previous matrix while providing additional insights into the distributions and complex relationships between weather parameters. It underscores the dataset's richness and potential for in-depth meteorological analysis.

> The scatter plot matrix of weather parameters reveals complex relationships and distributions within the dataset. Temperature and dewpoint show a strong positive linear relationship, while precipitation variables are concentrated near zero with occasional higher values. Cloud cover variables exhibit non-linear relationships and clustering at 0% and 100%. Solar radiation components are strongly interrelated and negatively associated with cloud cover. Surface pressure and wind speed display weak to moderate relationships with other variables, with interesting patterns. Wind direction is relatively uniformly distributed with slight directional preferences. Latitude and longitude show distinct clusters, indicating data collection from specific locations. The data block ID's lack of solid relationships with other variables suggests consistent data collection across different periods or locations. Overall, this visualization reinforces the correlations observed in the previous matrix while highlighting the intricate, often non-linear interactions between weather parameters, underscoring the dataset's richness and potential for comprehensive meteorological analysis.

# Machine Learning with Spark/MLlib

## Setup and Data *Loading*

> First, we set up the Spark session and loaded the weather data into a Spark DataFrame.

In [ ]:
!pip install pyspark

> The script code uses PySpark to set up a Spark session, load a historical weather dataset from a CSV file into a Spark DataFrame, and display the first few rows of the DataFrame to verify the successful loading of the data, which is typically the initial step in a data processing pipeline using PySpark before performing further operations such as data cleaning, transformation, and analysis.

In [ ]:
# Import SparkSession class from PySpark SQL module
from pyspark.sql import SparkSession

# Import col function from PySpark SQL functions module
from pyspark.sql.functions import col

# Start Spark session
spark = SparkSession.builder \
    .appName("Solar Energy Forecasting") \
    .getOrCreate()

# Load the dataset
file_path = data_path
df = spark.read.csv(file_path, header=True, inferSchema=True)

# Convert to Pandas DataFrame and transpose
pandas_df = df.limit(4).toPandas()
transposed_df = pandas_df.T

# Show the first few rows of the DataFrame to confirm successful load
print(transposed_df)


## Data Preprocessing

> Since no missing values are per the EDA, we will focus on feature scaling and encoding any categorical variables if necessary.

> The script code uses PySpark's machine learning library, specifically the `VectorAssembler` and `StandardScaler`, to preprocess the data by assembling the numerical features into a single vector column and then scaling those features to have zero mean and unit variance, which is a common technique in machine learning to ensure that all features have similar magnitudes and prevent certain features from dominating the model training process.

In [ ]:
# Import VectorAssembler and StandardScaler from PySpark ML feature module
from pyspark.ml.feature import VectorAssembler, StandardScaler

# scale numerical features 'datetime' and 'data_block_id' by encoded or extracted for features
numeric_features = [col_name for col_name in df.columns if col_name not in ['datetime', 'data_block_id']]
assembler = VectorAssembler(inputCols=numeric_features, outputCol="features_unscaled")
df_unscaled = assembler.transform(df)

# Scale features
scaler = StandardScaler(inputCol="features_unscaled", outputCol="features", withStd=True)
scaler_model = scaler.fit(df_unscaled)
df_scaled = scaler_model.transform(df_unscaled)

# Show transformed data
df_scaled.select("Features").show(5)


## Feature Engineering

> Create additional features that may help improve model performance, such as time-related features from the datetime column.

> The script code further preprocesses the data by extracting relevant date and time information from the "datetime" column using PySpark's built-in functions `month()`, `dayofmonth()`, and `hour()`, creating new columns for the month, day, and hour. These new features are then added to the list of feature columns, and the `VectorAssembler` is re-run to include these additional features in the final assembled feature vector, resulting in a DataFrame (`df_final`) that now contains the original numerical features along with the newly extracted date and time features, all combined into a single "features" column ready for use in machine learning models.

In [ ]:
# Import month, dayofmonth, and hour functions from PySpark SQL functions module
from pyspark.sql.functions import month, dayofmonth, hour

# Extracting date parts
df = df.withColumn("month", month("datetime"))
df = df.withColumn("day", dayofmonth("datetime"))
df = df.withColumn("hour", hour("datetime"))

# Re-run the assembler with the new features
new_feature_cols = numeric_features + ["month", "day", "hour"]
assembler = VectorAssembler(inputCols=new_feature_cols, outputCol="features")
df_final = assembler.transform(df)


## Building Machine Learning Models

> Utilize Spark MLlib to train a model, such as linear regression, to predict solar radiation based on the features.

### MLP (Multilayer Perceptron)

Multilayer Perceptrons (MLPs), also known as feedforward neural networks, are valuable for energy forecasting projects due to their versatility and ability to handle complex data relationships. MLPs excel at capturing non-linear interactions in energy systems where multiple factors interplay intricately. They offer flexibility in handling various input variables and can be used for regression and classification tasks in energy forecasting. MLPs automatically learn relevant features from input data, potentially uncovering hidden patterns. They are scalable and capable of processing large datasets and high-dimensional input spaces, making them suitable for complex energy forecasting scenarios. MLPs adapt to evolving consumption patterns and can be configured for multi-output prediction, allowing simultaneous forecasting of multiple energy variables. With proper preprocessing, they can handle datasets with missing values shared in real-world energy data. MLPs can be effectively combined with other models in ensemble methods to improve overall forecasting accuracy. They balance model complexity and interpretability compared to more advanced deep learning architectures. The architecture of MLPs is easily customizable to suit specific energy forecasting tasks. While implementing MLPs requires careful consideration of data preprocessing, architecture design, and hyperparameter tuning, they provide a powerful tool for capturing complex patterns in energy data, bridging the gap between simple statistical models and more advanced deep learning approaches.

> This script implements a Multilayer Perceptron (MLP) neural network using PySpark and TensorFlow to predict shortwave radiation based on historical weather data. It loads and preprocesses the data using PySpark, normalizes it, creates time series sequences, builds and trains an MLP model, generates predictions, and then post-processes the results. The script joins the predictions with the original data, performs data quality checks, and evaluates the model's performance using various regression metrics (RMSE, MAE, R2, MAPE). It leverages PySpark's distributed computing capabilities for data handling and TensorFlow for deep learning. It combines big data processing with neural network techniques to create a scalable solution for time series prediction on weather data. The script includes error handling and informative print statements to track progress and catch issues, concluding with a display of the final results, including the original data and the model's predictions.

In [ ]:
# Import necessary libraries
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, monotonically_increasing_id, isnan, when, abs as spark_abs, avg
from pyspark.sql.types import DoubleType, FloatType
from pyspark.ml.evaluation import RegressionEvaluator
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.optimizers import Adam
from sklearn.preprocessing import MinMaxScaler

# Initialize Spark session
spark = SparkSession.builder.appName("MLP with PySpark").getOrCreate()

def print_df_info(df, name):
    print(f"\n--- {name} Info ---")
    print(f"Count: {df.count()}")
    print("Schema:")
    df.printSchema()
    print("Sample data:")
    df.show(5, truncate=False)

try:
    # Step 1: Load and prepare data
    print("\nStep 1: Loading and preparing data...")
    df = spark.read.csv("/content/drive/MyDrive/big_data_project/historical_weather.csv", header=True, inferSchema=True)
    pandas_df = df.select(col('shortwave_radiation')).toPandas()
    data = pandas_df.values

    # Step 2: Normalize the data
    print("\nStep 2: Normalizing data...")
    scaler = MinMaxScaler(feature_range=(0, 1))
    scaled_data = scaler.fit_transform(data)

    # Step 3: Prepare input/output sequences
    print("\nStep 3: Preparing input/output sequences...")
    X, y = [], []
    sequence_length = 10  # Number of time steps to look back
    for i in range(len(scaled_data) - sequence_length):
        X.append(scaled_data[i:i+sequence_length])
        y.append(scaled_data[i+sequence_length])
    X, y = np.array(X), np.array(y)

    # Step 4: Build MLP model
    print("\nStep 4: Building MLP model...")
    model = Sequential([
        Dense(64, activation='relu', input_shape=(sequence_length, 1)),
        Dropout(0.2),
        Dense(32, activation='relu'),
        Dropout(0.2),
        Dense(16, activation='relu'),
        Dense(1)
    ])
    model.compile(optimizer=Adam(learning_rate=0.001), loss='mse')

    # Step 5: Train the model
    print("\nStep 5: Training the model...")
    model.fit(X, y, epochs=10, batch_size=32, validation_split=0.2, verbose=1)

    # Step 6: Make predictions
    print("\nStep 6: Making predictions...")
    predictions = model.predict(X)
    print(f"Predictions shape: {predictions.shape}")
    print(f"Predictions sample: {predictions[:5]}")

    # Step 7: Convert predictions back to original scale
    print("\nStep 7: Converting predictions to original scale...")
    # Reshape predictions to 2D array
    predictions_2d = predictions.reshape(-1, 1)
    predictions_original = scaler.inverse_transform(predictions_2d)
    print(f"Inverse transformed predictions shape: {predictions_original.shape}")
    print(f"Inverse transformed predictions sample: {predictions_original[:5]}")

    # Step 8: Convert predictions to Spark DataFrame
    print("\nStep 8: Converting predictions to Spark DataFrame...")
    predictions_df = spark.createDataFrame(predictions_original.tolist(), ["prediction"])
    print_df_info(predictions_df, "Predictions DataFrame")

    # Step 9: Join with original data
    print("\nStep 9: Joining predictions with original data...")
    print_df_info(df, "Original DataFrame")

    # Adjust the number of rows in predictions_df to match the original df
    predictions_df = predictions_df.limit(df.count() - sequence_length)

    result_df = df.withColumn("row_id", monotonically_increasing_id()).join(
        predictions_df.withColumn("row_id", monotonically_increasing_id()),
        "row_id"
    ).drop("row_id")
    print_df_info(result_df, "Result DataFrame")

    # Check for NaN or null values
    print("\nChecking for NaN or null values...")
    for column in result_df.columns:
        null_count = result_df.filter(col(column).isNull()).count()
        if result_df.schema[column].dataType in [DoubleType(), FloatType()]:
            nan_count = result_df.filter(isnan(col(column))).count()
            print(f"Column '{column}': Null count = {null_count}, NaN count = {nan_count}")
        else:
            print(f"Column '{column}': Null count = {null_count}")

    # Step 10: Evaluate the model
    print("\nStep 10: Evaluating the model...")
    evaluator = RegressionEvaluator(labelCol="shortwave_radiation", predictionCol="prediction")

    # Ensure 'shortwave_radiation' and 'prediction' columns exist and are of the correct type
    if "shortwave_radiation" not in result_df.columns or "prediction" not in result_df.columns:
        raise ValueError("'shortwave_radiation' or 'prediction' column missing from result_df")

    result_df = result_df.withColumn("shortwave_radiation", col("shortwave_radiation").cast("double"))
    result_df = result_df.withColumn("prediction", col("prediction").cast("double"))

    # Remove rows with NaN or null values in relevant columns
    result_df = result_df.filter(
        ~isnan(col("shortwave_radiation")) & ~isnan(col("prediction")) &
        col("shortwave_radiation").isNotNull() & col("prediction").isNotNull()
    )

    # Calculate metrics
    metrics = ["rmse", "mae", "r2"]
    for metric in metrics:
        value = evaluator.evaluate(result_df, {evaluator.metricName: metric})
        print(f"{metric.upper()}: {value}")

    # Calculate MAPE
    result_df = result_df.withColumn("mape",
        when(col("shortwave_radiation") != 0,
             spark_abs(col("shortwave_radiation") - col("prediction")) / col("shortwave_radiation") * 100)
        .otherwise(None)
    ).withColumn("mape", col("mape").cast("double"))  # Ensure MAPE is cast to double

    # Calculate average MAPE using Spark SQL functions
    mape = result_df.select(avg("mape")).collect()[0][0]
    print(f"MAPE: {mape}%")

    # Step 11: Show the results
    print("\nStep 11: Showing final results...")
    result_df.show()

except Exception as e:
    print(f"An error occurred: {str(e)}")
    import traceback
    print(traceback.format_exc())

finally:
    # Stop the Spark session
    print("\nStopping Spark session...")
    spark.stop()


#### Model Evaluation

> Key findings and analysis of the model:

1. Model Architecture:
   The model uses a Multi-Layer Perceptron (MLP) neural network implemented with TensorFlow/Keras. It has three dense layers with ReLU activation and dropout layers for regularization.

2. Data Preparation:
   - The data is loaded from a CSV file containing historical weather data.
   - The 'shortwave_radiation' column is used as the target variable.
   - The data is normalized using MinMaxScaler to scale values between 0 and 1.
   - A sequence length of 10 time steps is used for input features.

3. Training Process:
   - The model was trained for ten epochs with a batch size 32.
   - The training loss decreased slightly from 0.0086 to 0.0080 over the epochs.
   - The validation loss fluctuated between 0.0140 and 0.0171, indicating some overfitting.

4. Prediction Results:
   - The model made predictions on the test set, which were then inverse-transformed to the original scale.
   - The predictions were joined with the original data for evaluation.

5. Model Performance:
   The evaluation metrics show poor performance:
   - RMSE (Root Mean Square Error): 215.1385851080044
   - MAE (Mean Absolute Error): 142.19445388320852
   - R2 (R-squared): -0.3341610588554853
   - MAPE (Mean Absolute Percentage Error): 436.75917882994685%

6. Interpretation of Results:
   - The negative R2 score indicates that the model performs worse than a horizontal line (i.e., predicting the mean of the target variable).
   - The high MAPE (436.76%) suggests that the model's predictions are, on average, off by more than four times the actual values.
   - The RMSE and MAE values are large, indicating significant prediction errors.

7. Data Quality:
   - The final dataset has no null or NaN values, which is positive for data quality.
   - The 'shortwave_radiation' column (target variable) contains many zero values, which may be challenging for the model to predict accurately.

8. Potential Issues:
   - The model is overfitting, as indicated by the divergence between training and validation loss.
   - The poor performance metrics suggest that the current model architecture or feature set may not be suitable for this prediction task.
   - The high number of zero values in the target variable ('shortwave_radiation') may affect the model's ability to learn meaningful patterns.

Recommendations:
1. Feature engineering: Consider incorporating more relevant features or creating derived features that might better correlate with shortwave radiation.
2. Model architecture: Experiment with different neural network architectures, such as LSTM or GRU layers, which might be more suitable for time series data.
3. Hyperparameter tuning: Use techniques like grid search or random search to optimize hyperparameters.
4. Handle zero values: Investigate the zero values in the target variable and consider strategies to handle them, such as separate models for zero/non-zero prediction.
5. Ensemble methods: To compare performance, try ensemble techniques or other machine learning algorithms (e.g., Random Forests, Gradient Boosting).
6. Longer training: Increase the number of epochs and implement early stopping to prevent overfitting while allowing for more extended training.
7. Data augmentation: Gather more data or use data augmentation techniques to increase the dataset size.

In conclusion, while the current model implementation shows poor performance, several avenues exist for improvement. Predicting shortwave radiation based on historical weather data is complex and requires more sophisticated modelling techniques or additional feature engineering to achieve satisfactory results.

#### Hyperparameter Tuning

In [ ]:
# Import necessary libraries
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping

def print_df_info(df, name):
    print(f"\n--- {name} Info ---")
    print(f"Shape: {df.shape}")
    print("Sample data:")
    print(df.head())

def main():
    try:
        # Step 1: Load and prepare data
        print("\nStep 1: Loading and preparing data...")
        file_path = "/content/drive/MyDrive/big_data_project/historical_weather.csv"
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"The file {file_path} does not exist. Please check the path.")

        # Read only 100k records
        df = pd.read_csv(file_path, nrows=100000)
        print_df_info(df, "Original DataFrame")

        # Feature engineering
        df['datetime'] = pd.to_datetime(df['datetime'])
        df['prev_shortwave_radiation'] = df['shortwave_radiation'].shift(1)
        df['radiation_change'] = df['shortwave_radiation'] - df['prev_shortwave_radiation']
        df['day_of_year'] = df['datetime'].dt.dayofyear
        df['hour'] = df['datetime'].dt.hour

        # Select features for prediction
        feature_cols = ["temperature", "dewpoint", "rain", "snowfall", "surface_pressure",
                        "cloudcover_total", "windspeed_10m", "prev_shortwave_radiation",
                        "radiation_change", "day_of_year", "hour", "latitude", "longitude"]

        # Handle null values
        df[feature_cols + ["shortwave_radiation"]] = df[feature_cols + ["shortwave_radiation"]].fillna(0)

        print_df_info(df[feature_cols + ["shortwave_radiation"]], "Processed DataFrame")

        # Prepare data for ML
        X = df[feature_cols]
        y = df["shortwave_radiation"]

        # Split data
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

        # Step 2: Train Random Forest model
        print("\nStep 2: Training Random Forest model...")
        rf = RandomForestRegressor(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1)
        rf.fit(X_train, y_train)

        # Make predictions
        rf_predictions = rf.predict(X_test)

        # Evaluate Random Forest model
        rf_rmse = np.sqrt(mean_squared_error(y_test, rf_predictions))
        print(f"Random Forest RMSE: {rf_rmse}")

        # Step 3: Prepare data for neural network
        print("\nStep 3: Preparing data for neural network...")
        scaler = MinMaxScaler(feature_range=(0, 1))
        X_scaled = scaler.fit_transform(X)
        y_scaled = scaler.fit_transform(y.values.reshape(-1, 1))

        # Prepare input/output sequences
        sequence_length = 24
        X_seq, y_seq = [], []
        for i in range(len(X_scaled) - sequence_length):
            X_seq.append(X_scaled[i:i+sequence_length])
            y_seq.append(y_scaled[i+sequence_length])
        X_seq, y_seq = np.array(X_seq), np.array(y_seq)

        # Split the data
        X_train, X_test, y_train, y_test = train_test_split(X_seq, y_seq, test_size=0.2, random_state=42)

        # Step 4: Build neural network model
        print("\nStep 4: Building neural network model...")
        model = Sequential([
            LSTM(64, activation='relu', input_shape=(sequence_length, len(feature_cols))),
            Dense(32, activation='relu'),
            Dense(1)
        ])
        model.compile(optimizer=Adam(learning_rate=0.001), loss='mse')

        # Step 5: Train the model with early stopping
        print("\nStep 5: Training the model...")
        early_stopping = EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True)
        history = model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2, callbacks=[early_stopping], verbose=1)

        # Step 6: Make predictions
        print("\nStep 6: Making predictions...")
        predictions = model.predict(X_test)

        # Step 7: Convert predictions back to original scale
        print("\nStep 7: Converting predictions to original scale...")
        predictions_original = scaler.inverse_transform(predictions)
        y_test_original = scaler.inverse_transform(y_test)

        # Step 8: Evaluate the model
        print("\nStep 8: Evaluating the model...")
        mse = mean_squared_error(y_test_original, predictions_original)
        rmse = np.sqrt(mse)
        mae = mean_absolute_error(y_test_original, predictions_original)
        r2 = r2_score(y_test_original, predictions_original)

        print(f"RMSE: {rmse}")
        print(f"MAE: {mae}")
        print(f"R2 Score: {r2}")

        # Step 9: Feature importance from Random Forest
        print("\nStep 9: Feature importance from Random Forest...")
        feature_importance = rf.feature_importances_
        for feature, importance in zip(feature_cols, feature_importance):
            print(f"{feature}: {importance}")

    except Exception as e:
        print(f"An error occurred: {str(e)}")
        import traceback
        print(traceback.format_exc())

if __name__ == "__main__":
    main()


> This code captures the script's end time, formats timestamps for both start and end times, calculates the total execution time and prints a summary including start time, end time, and duration in a structured format.

In [ ]:
import time
from datetime import timedelta

end_time = time.time()
end_time_str = time.strftime('%Y-%m-%d %H:%M:%S')
execution_time = end_time - start_time

# Convert execution time to hours, minutes, seconds
time_delta = timedelta(seconds=int(execution_time))
hours, remainder = divmod(time_delta.seconds, 3600)
minutes, seconds = divmod(remainder, 60)

print(f"\nScript execution summary:")
print(f"Started at..: {start_time_str}")
print(f"Ended at....: {end_time_str}")
print(f"Total execution time: {hours} hours, {minutes} minutes, and {seconds} seconds")
